# Project 1a
brute force approach of finding all substitutions in a sequence of genomes compared to a reference genome

# Project 1b
Using a hash table to find all substitutions, insertions, and deletions within a sequence of genomes compared to a reference genome

# Project 1c
Using smith-waterman algorithm as well as hash tables to figure out which sequence of DNA belongs to which genome. Comparing 100 different reference genomes that the DNA could potentially belong to. 

# Project 1d
Using smith-waterman algorithm as well as hash tables to figure out which sequence of DNA belongs to which genome. Comparing 5000 different reference genomes that the DNA could potentially belong to. 

# Project 2a
Using a de Bruijn graph to reconstruct a genome using perfect sequences of DNA (no errors)

# Project 2b
Using a de Bruijn graph to reconstruct a genome using imperfect sequences of DNA. Used de Bruijn graph method to construct the longest path out of these sequences.

# Project 3a
Finding the most probable motif kmers within a series of sequences using Probability Weight Matrices (PWMs) and a greedy algorithm motif search. 

# Project 3b
Finding the top sequences in which a motif might occur using PWMs and a greedy algorithm motif search. 
